### ISSUE TITLE
ECSV example linked from the documention fails with `RepresenterError`

### ISSUE THREAD

**Issue by hamogu** [astropy/astropy#6156]
The astropy docs for ECSV link to [APE 6](https://github.com/astropy/astropy-APEs/blob/master/APE6.rst) where the ECSV format was specified. However, I cannot actually write the implementation example given in the APE.

If the ECSV format definiton has changed in a non-backwards compatible way, we should link to the updated format definition.

```
melkor:~> ipython
Python 2.7.12 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:42:40) 
Type "copyright", "credits" or "license" for more information.

IPython 5.3.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from astropy.utils import OrderedDict

In [3]: from astropy.table import Table

In [4]: t = Table([[1, 4], [2, 3]], names=['a', 'b'])

In [5]: t.meta['keywords'] = OrderedDict([('z_key1', 'val1'), ('a_key2', 'val2')
   ...: ])
WARNING: AstropyDeprecationWarning: astropy.utils.compat.odict.OrderedDict is now deprecated - import OrderedDict from the collections module instead [astropy.utils.compat.odict]

In [6]: t.meta['comments'] = ['Comment 1', 'Comment 2', 'Comment 3']

In [8]: from astropy import units as u

In [9]: t['a'].unit = u.m / u.s

In [10]: t['a'].format = '%5.2f'

In [11]: t['a'].description = 'Column A'

In [12]: t['b'].meta = dict(column_meta={'a':1, 'b': 2})

In [13]: import sys

In [14]: t.write(sys.stdout, format='ascii.ecsv')
---------------------------------------------------------------------------
RepresenterError                          Traceback (most recent call last)
<ipython-input-14-fee9bce7935a> in <module>()
----> 1 t.write(sys.stdout, format='ascii.ecsv')

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/table.pyc in write(self, *args, **kwargs)
   2445         passed through to the underlying data reader (e.g. `~astropy.io.ascii.write`).
   2446         """
-> 2447         io_registry.write(self, *args, **kwargs)
   2448 
   2449     def copy(self, copy_data=True):

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/registry.pyc in write(data, *args, **kwargs)
    516 
    517     writer = get_writer(format, data.__class__)
--> 518     writer(data, *args, **kwargs)
    519 
    520 

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/ascii/connect.pyc in io_write(format, table, filename, **kwargs)
     41     from .ui import write
     42     format = re.sub(r'^ascii\.', '', format)
---> 43     return write(table, filename, format=format, **kwargs)
     44 
     45 

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/ascii/ui.pyc in write(table, output, format, Writer, fast_writer, **kwargs)
    727         return
    728 
--> 729     lines = writer.write(table)
    730 
    731     # Write the lines to output

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/ascii/core.pyc in write(self, table)
   1267         # Write header and data to lines list
   1268         lines = []
-> 1269         self.write_header(lines, table.meta)
   1270         self.data.write(lines)
   1271 

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/ascii/core.pyc in write_header(self, lines, meta)
   1234     def write_header(self, lines, meta):
   1235         self.header.write_comments(lines, meta)
-> 1236         self.header.write(lines)
   1237 
   1238     def write(self, table):

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/io/ascii/ecsv.pyc in write(self, lines)
     70         header_yaml_lines = (['%ECSV {0}'.format(ECSV_VERSION),
     71                               '---']
---> 72                              + meta.get_yaml_from_header(header))
     73 
     74         lines.extend([self.write_comment + line for line in header_yaml_lines])

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/meta.pyc in get_yaml_from_header(header)
    284     del header['cols']
    285 
--> 286     lines = yaml.dump(header, Dumper=TableDumper).splitlines()
    287     return lines
    288 

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/__init__.pyc in dump(data, stream, Dumper, **kwds)
    200     If stream is None, return the produced string instead.
    201     """
--> 202     return dump_all([data], stream, Dumper=Dumper, **kwds)
    203 
    204 def safe_dump_all(documents, stream=None, **kwds):

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/__init__.pyc in dump_all(documents, stream, Dumper, default_style, default_flow_style, canonical, indent, width, allow_unicode, line_break, encoding, explicit_start, explicit_end, version, tags)
    188         dumper.open()
    189         for data in documents:
--> 190             dumper.represent(data)
    191         dumper.close()
    192     finally:

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent(self, data)
     26 
     27     def represent(self, data):
---> 28         node = self.represent_data(data)
     29         self.serialize(node)
     30         self.represented_objects = {}

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_data(self, data)
     55             data_types = self.get_classobj_bases(data.__class__)+list(data_types)
     56         if data_types[0] in self.yaml_representers:
---> 57             node = self.yaml_representers[data_types[0]](self, data)
     58         else:
     59             for data_type in data_types:

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_dict(self, data)
    223 
    224     def represent_dict(self, data):
--> 225         return self.represent_mapping(u'tag:yaml.org,2002:map', data)
    226 
    227     def represent_set(self, data):

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/meta.pyc in represent_mapping(self, tag, mapping, flow_style)
    264             for item_key, item_value in mapping:
    265                 node_key = self.represent_data(item_key)
--> 266                 node_value = self.represent_data(item_value)
    267                 if not (isinstance(node_key, yaml.ScalarNode) and not node_key.style):
    268                     best_style = False

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_data(self, data)
     55             data_types = self.get_classobj_bases(data.__class__)+list(data_types)
     56         if data_types[0] in self.yaml_representers:
---> 57             node = self.yaml_representers[data_types[0]](self, data)
     58         else:
     59             for data_type in data_types:

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/meta.pyc in _repr_odict(dumper, data)
    142     '!!omap [foo: bar, mumble: quux, baz: gorp]\\n'
    143     """
--> 144     return _repr_pairs(dumper, u'tag:yaml.org,2002:omap', six.iteritems(data))
    145 
    146 

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/meta.pyc in _repr_pairs(dump, tag, sequence, flow_style)
    117     best_style = True
    118     for (key, val) in sequence:
--> 119         item = dump.represent_data({key: val})
    120         if not (isinstance(item, yaml.ScalarNode) and not item.style):
    121             best_style = False

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_data(self, data)
     55             data_types = self.get_classobj_bases(data.__class__)+list(data_types)
     56         if data_types[0] in self.yaml_representers:
---> 57             node = self.yaml_representers[data_types[0]](self, data)
     58         else:
     59             for data_type in data_types:

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_dict(self, data)
    223 
    224     def represent_dict(self, data):
--> 225         return self.represent_mapping(u'tag:yaml.org,2002:map', data)
    226 
    227     def represent_set(self, data):

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/astropy/table/meta.pyc in represent_mapping(self, tag, mapping, flow_style)
    264             for item_key, item_value in mapping:
    265                 node_key = self.represent_data(item_key)
--> 266                 node_value = self.represent_data(item_value)
    267                 if not (isinstance(node_key, yaml.ScalarNode) and not node_key.style):
    268                     best_style = False

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_data(self, data)
     65                     node = self.yaml_multi_representers[None](self, data)
     66                 elif None in self.yaml_representers:
---> 67                     node = self.yaml_representers[None](self, data)
     68                 else:
     69                     node = ScalarNode(None, unicode(data))

/nfs/melkor/d1/guenther/soft/anaconda/lib/python2.7/site-packages/yaml/representer.pyc in represent_undefined(self, data)
    247 
    248     def represent_undefined(self, data):
--> 249         raise RepresenterError("cannot represent an object: %s" % data)
    250 
    251 SafeRepresenter.add_representer(type(None),

RepresenterError: cannot represent an object: OrderedDict([('z_key1', 'val1'), ('a_key2', 'val2')])

In [46]: yaml.__version__
Out[46]: '3.12'
```

That's apparently the current version with anadonca for python 2.7 (although it's labelled differently in conda ?!?):

```
melkor:~> conda upgrade yaml
Fetching package metadata ...........
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /nfs/melkor/d1/guenther/soft/anaconda:
#
yaml                      0.1.6                         0  
```

@taldcroft

**Comment by taldcroft** [comment]
I can indeed reproduce this, but the example works for me with `collections.OrderedDict`.  The astropy compatibility version of `OrderedDict` is deprecated since we no longer support Python 2.6.  I'm surprised about the hard crash, but the immediate fix is `from collections import OrderedDict` to and to put in a PR to fix this in APE6.  I will do this right now.

**Comment by hamogu** [comment]
Since APEs live in a different repro, and it's a problem there and not here, I'll close this.

**Comment by hamogu** [comment]
I only just realized the the default format for ecsv is actually space limited, not "comma -separated values". I understand the heritage, but given the name of the format that's surprising (I did not even think to check, I just noticed when loading a hand-written files with commas in it failed.
Not sure it's worth a new issue, but I would suggest to highlight this fact in the docs or at least in the APE that the docs link to.
